### PR TITLE
OCPBUGS-13830: Add support for overriding container resources

### PR DESCRIPTION
**Pull request by arkadeepsen**
This PR adds the OverrideResources field to DeploymentConfig. This would add the support of overriding the resources fields of the cert-manager, the cert-manager-cainjector, and the cert-manager-webhook deploymemts. Additionally, the resource requests for cert-manager-operator deployment is added.

### CODE REVIEW

**Comment by TrilokGeer** [comment]
Can PR description and subject be updated? Maybe https://hypershift-docs.netlify.app/contribute/ is of help.

File: pkg/controller/deployment/deployment_helper.go [hunk 1]
```diff
@@ -76,6 +76,74 @@ func parseArgMap(argMap map[string]string, args []string) {
...
+// provided as override resources.
+func mergeContainerResources(sourceResources corev1.ResourceRequirements, overrideResources corev1.ResourceRequirements) corev1.ResourceRequirements {
+	mergedResourceLimits := mergeContainerResourceList(sourceResources.Limits, overrideResources.Limits)
+	mergedResourceRequests := mergeContainerResourceList(sourceResources.Requests, overrideResources.Requests)
+	mergedResourceClaims := mergeResourceClaims(sourceResources.Claims, overrideResources.Claims)
+
+	return corev1.ResourceRequirements{
+		Limits:   mergedResourceLimits,
+		Requests: mergedResourceRequests,
+		Claims:   mergedResourceClaims,
```
[the remark this answers is not in the document]
Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_helper.go:89]: Makes sense. I had thought as the field is included, better to support the merge functionality. Will remove it.

File: pkg/controller/deployment/deployment_helper.go [hunk 2]
```diff
@@ -76,6 +76,74 @@ func parseArgMap(argMap map[string]string, args []string) {
 	}
 }
 
+// mergeContainerResources merges source container resources with that
+// provided as override resources.
+func mergeContainerResources(sourceResources corev1.ResourceRequirements, overrideResources corev1.ResourceRequirements) corev1.ResourceRequirements {
```
Comment (thejasn) [inline pkg/controller/deployment/deployment_helper.go:81]: Is it possible to make use of https://github.com/kubernetes/kubernetes/blob/7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647/pkg/api/v1/resource/helpers.go#L377?

Comment (swghosh) [reply pkg/controller/deployment/deployment_helper.go:81]: +1

Comment (swghosh) [reply pkg/controller/deployment/deployment_helper.go:81]: The func signature in this `MergeContainerResourceLimits` fn uses `v1.Container` type and two things to note:
- it doesn't set `hugepage` related resources
- it doesn't set the overriden limit in case the limit was already set in the container (i.e. if we use this and in the future we end up adding resource *limits* to our operator deployment (which is currently not present) we have to change this implementation)

File: bundle/manifests/cert-manager-operator.clusterserviceversion.yaml [hunk 3]
```diff
@@ -664,7 +664,10 @@ spec:
                 image: openshift.io/cert-manager-operator:latest
                 imagePullPolicy: Always
                 name: cert-manager-operator
-                resources: {}
+                resources:
```
[the remark this answers is not in the document]
Comment (swghosh) [reply bundle/manifests/cert-manager-operator.clusterserviceversion.yaml:667]: Is the operator container resource requests/limits customizable in any way?
We're setting it to these values here in the CSV which is a default. Asking this because, the original bug report is also about the operator resources and not just for operand.

File: pkg/controller/deployment/deployment_helper.go [hunk 4]
```diff
@@ -76,6 +76,74 @@ func parseArgMap(argMap map[string]string, args []string) {
...
+func mergeContainerResourceList(sourceResourceList corev1.ResourceList, overrideResourceList corev1.ResourceList) corev1.ResourceList {
+	if sourceResourceList == nil && overrideResourceList == nil {
+		return nil
+	}
+
+	destResourceList := corev1.ResourceList{}
+	parseResourceList(destResourceList, sourceResourceList)
+	parseResourceList(destResourceList, overrideResourceList)
+
+	return destResourceList
```
Comment (thejasn) [inline pkg/controller/deployment/deployment_helper.go:104]: This feels redundant as `ResourceList` is already a mad[resourceName]Quantity. Why can't we just explicitly set values for the `Memory` and `CPU`?

File: pkg/controller/deployment/deployment_helper.go (hunk 2 above)
Comment (thejasn) [reply pkg/controller/deployment/deployment_helper.go:81]: signature change shouldn't matter to us, but anyways we can't reuse. 

> it doesn't set hugepage related resources
We do not want to support hugepages either since this will involve overriding volumes and volumemounts. So until someone files a bug/rfe. 

> it doesn't set the overriden limit in case the limit was already set in the container (i.e. if we use this and in the future we end up adding resource limits to our operand deployment (which is currently not present) we have to change this implementation)
Yeah, this might be an issue.

File: pkg/controller/deployment/deployment_helper.go (hunk 4 above)
Comment (thejasn) [reply pkg/controller/deployment/deployment_helper.go:104]:
```suggestion
	if sourceResourceList == nil && overrideResourceList == nil {
		return nil
	}

	destResourceList := corev1.ResourceList{}
	if !sourceResourceList[corev1.ResourceCPU].IsZero() {
	        destResourceList[corev1.ResourceCPU] = sourceResourceList[corev1.ResourceCPU].DeepCopy()
	}
	
	if !sourceResourceList[corev1.ResourceMemory].IsZero() {
	        destResourceList[corev1.ResourceMemory] = sourceResourceList[corev1.ResourceMemory].DeepCopy()
	}

	return destResourceList
```

File: pkg/controller/deployment/deployment_helper.go [hunk 5]
```diff
@@ -76,6 +76,74 @@ func parseArgMap(argMap map[string]string, args []string) {
...
+	return corev1.ResourceRequirements{
+		Limits:   mergedResourceLimits,
+		Requests: mergedResourceRequests,
+		Claims:   mergedResourceClaims,
+	}
+}
+
+// mergeContainerResourceList merges source resource list with that
+// provided as override resource list.
+func mergeContainerResourceList(sourceResourceList corev1.ResourceList, overrideResourceList corev1.ResourceList) corev1.ResourceList {
```
Comment (thejasn) [inline pkg/controller/deployment/deployment_helper.go:95]: Technically not a merge in case of `ResourceList` since we are basically overriding only here. So might as well rename this.

Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_helper.go:95]: Won't we merge the overriding resources with the source's resources if, say, only limits is provided in the overriding resources and not requests, whereas source's resources has both limits and requests?

If we override then whatever is in the overriding resources, even if either of empty limits or requests, will override the source's limits and requests even if something is already set there.

Comment (thejasn) [reply pkg/controller/deployment/deployment_helper.go:95]: Overriding resources takes priority so we set whatever is set there as the final value. What you are suggesting would be the case if one could update source's resources via a different channel as compared `overrideResources`.

File: api/operator/v1alpha1/certmanager_types.go [hunk 6]
```diff
@@ -77,6 +77,10 @@ type DeploymentConfig struct {
 	// +kubebuilder:validation:Optional
 	// +optional
 	OverrideLabels map[string]string `json:"overrideLabels,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +optional
+	OverrideResources corev1.ResourceRequirements `json:"overrideResources,omitempty"`
```
Comment (swghosh) [inline api/operator/v1alpha1/certmanager_types.go:83]: Can we change this type to use a self-created struct that just contains Resources and Limits (essentially stripping out Claims as well, which would be counter-intuitive for discouraging users about adding any Claims as well).

It's better to not rely upon upstream type, rather have a copied one with only the necessary fields, xref: https://github.com/openshift/api/pull/1387#discussion_r1182817836

File: pkg/controller/deployment/deployment_helper.go (hunk 5 above)
Comment (thejasn) [reply pkg/controller/deployment/deployment_helper.go:95]: We will never have default resources, so yes. And if `overrideResources` is `nil` nothing to do but if `overrideResources` is `{}` then yes that will take precedence.

Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_helper.go:95]: I had a discussion with @TrilokGeer that we might want to do some study and set some default values in the future. For reference IBM's cert-manager operator sets some default values: https://github.com/IBM/ibm-cert-manager-operator/blob/master/controllers/resources/certmanager.go

File: bundle/manifests/cert-manager-operator.clusterserviceversion.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (thejasn) [reply bundle/manifests/cert-manager-operator.clusterserviceversion.yaml:667]: We are only setting the `requests` so should be ok, but yes limits need to be customizable. Can be done using `subscription.spec.config` so not an issue.

File: pkg/controller/deployment/deployment_helper.go (hunk 5 above)
Comment (thejasn) [reply pkg/controller/deployment/deployment_helper.go:95]: For requests we could have something in the future but for limits I'd still be hesitant to push during upgrades especially when majority of the users are using `Automatic` upgrades. This can be viewed as a potential breaking change during upgrades since we will not definitively know at what scale users are using cert-manager.  Limits are usually [frowned upon ](https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#resources-and-limits)on control plane components and although cert-manager doesn't affect control plane now it could in the future.

Comment (thejasn) [reply pkg/controller/deployment/deployment_helper.go:95]: Now that there is a possibility of adding default `requests` at least, I think we should be okay doing a `nil`/`empty` check before overriding so that we don't override with a nil/empty value for mem/cpu.

Comment (swghosh) [reply pkg/controller/deployment/deployment_helper.go:95]: > Limits are usually [frowned upon ](https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#resources-and-limits)on control plane components and although cert-manager doesn't affect control plane now it could in the future.

In the future, it is possible that API server endpoints use a cert-manager issued certificate and in that case if the operator/cert-manager operand goes down it could be a SLO problem.  I'd suggest preferring to refrain from adding any resource limits as of now (unless user themselves configures as such: which with this implementation they can do through overrideResources for operand and through Subscription object for the operator), also partly because upstream cert-manager also doesn't enforce these limits yet so we don't know what are good enough sane values for clusters with large certificate usage.

File: bundle/manifests/cert-manager-operator.clusterserviceversion.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (swghosh) [reply bundle/manifests/cert-manager-operator.clusterserviceversion.yaml:667]: Ok, we need to document it later (like we did for logLevels).
+doc-ack, cc: @xenolinux

File: pkg/controller/deployment/deployment_helper.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (swghosh) [reply pkg/controller/deployment/deployment_helper.go:89]: Added a follow-up comment in https://github.com/openshift/cert-manager-operator/pull/136#discussion_r1280991062.

File: pkg/controller/deployment/deployment_helper.go (hunk 5 above)
Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_helper.go:95]: So, if we set default requests and not limits, however, overriding resources has limits and no requests, do we remove the default requests and use whatever is in the overriding resources? Or do we keep the default resources and use the overriding limits?

Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_helper.go:95]: > In the future, it is possible that API server endpoints use a cert-manager issued certificate and in that case if the operator/cert-manager operand goes down it could be a SLO problem. I'd suggest preferring to refrain from adding any resource limits as of now (unless user themselves configures as such: which with this implementation they can do through overrideResources for operand and through Subscription object for the operator), also partly because upstream cert-manager also doesn't enforce these limits yet so we don't know what are good enough sane values for clusters with large certificate usage.

We might not set it right now, but the implementation should take into account the changes we might have in the future. So if we decide to add some default requests, an empty `overrideResources` will always end up removing that if we decide that it will override and not merge with default resources.

Comment (swghosh) [reply pkg/controller/deployment/deployment_helper.go:95]: Agreed, implementation should be in a way that it takes care of such scenarios.

I would say,
```yaml
originalDeployment:
  resources:
      requests: X
      limits: nil

overrides:
   resources:
     resources: nil
     limits: Y
```

should translate to:
```yaml
finalDeployment:
   resources:
     requests: X
     limits: Y
```
Hope it makes sense.

Comment (swghosh) [reply pkg/controller/deployment/deployment_helper.go:95]: Whereas,
```yaml
originalDeployment:
  resources:
      requests: X
      limits: nil

overrides:
   resources:
     resources: Z
     limits: Y
```

should still translate to:
```yaml
finalDeployment:
   resources:
     requests: Z
     limits: Y
```

Comment (swghosh) [reply pkg/controller/deployment/deployment_helper.go:95]: I was meaning to say, we stick to:
```yaml
finalDeployment:
  resources:
      requests:
        cpu: X
        memory: Y
      limits: nil
```
in such scenarios (which involves merging individual resource types, `iff not nil`).

File: pkg/controller/deployment/deploment_helper_test.go [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+							Limits: corev1.ResourceList{
+								corev1.ResourceCPU: k8sresource.MustParse("10m"),
+							},
+							Requests: corev1.ResourceList{
+								corev1.ResourceMemory: k8sresource.MustParse("128Mi"),
+							},
+						},
+					},
+				},
+			},
```
Comment (swghosh) [inline pkg/controller/deployment/deploment_helper_test.go:298]: Can a add another or a couple test case(s) to test the behaviour we discussed in case of:  https://github.com/openshift/cert-manager-operator/pull/136#discussion_r1281566526

Comment (arkadeepsen) [reply pkg/controller/deployment/deploment_helper_test.go:298]: Those are added here: https://github.com/openshift/cert-manager-operator/pull/136/files#diff-fe3b58419cf80f35e0dc128fc74a9aa700807ea529635eafa1e44c8d1f39ebd6R23

File: pkg/controller/deployment/deploment_helper_test.go [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+				},
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    k8sresource.MustParse("40m"),
+					corev1.ResourceMemory: k8sresource.MustParse("64Mi"),
+				},
+			},
+		},
+		{
+			name: "override resources limits and requests merges with source resource limits and requests respectively",
+			sourceResources: corev1.ResourceRequirements{
```
Comment (swghosh) [inline pkg/controller/deployment/deploment_helper_test.go:178]: Can you add test cases for testing limits and request as nil explicitly?

Similar to:
```
{
			name: "override resources limits and requests merges with source resource limits and requests respectively",
			sourceResources: corev1.ResourceRequirements{
				Limits: nil,
				Requests: corev1.ResourceList{
					corev1.ResourceCPU: k8sresource.MustParse("10m"),
				},
			},
			overrideResources: v1alpha1.CertManagerResourceRequirements{
				Limits: corev1.ResourceList{
					corev1.ResourceCPU: k8sresource.MustParse("400m"),
				},
				Requests: corev1.ResourceList{
					corev1.ResourceMemory: k8sresource.MustParse("64Mi"),
				},
			},
			expected: corev1.ResourceRequirements{
				Limits: corev1.ResourceList{
					corev1.ResourceCPU: k8sresource.MustParse("400m"),
				},
				Requests: corev1.ResourceList{
					corev1.ResourceCPU:    k8sresource.MustParse("10m"),
					corev1.ResourceMemory: k8sresource.MustParse("64Mi"),
				},
			},
		}
```

File: pkg/controller/deployment/deployment_overrides_test.go [hunk 9]
```diff
@@ -396,3 +400,145 @@ func TestParseArgMap(t *testing.T) {
...
+							Containers: []corev1.Container{
+								{
+									Name: "test-container",
+								},
+							},
+						},
+					},
+				},
+			},
+			overrideFn: func(_ certmanagerinformer.CertManagerInformer, _ string) (v1alpha1.CertManagerResourceRequirements, error) {
```
Comment (thejasn) [inline pkg/controller/deployment/deployment_overrides_test.go:430]: @arkadeepsen I don't the value running `withContainerResourcesOverrideHook()` in the test if are also passing the override test function. We can ditch this in favour of covering this as part of e2e.

File: pkg/controller/deployment/deployment_overrides_validation_test.go [hunk 10]
```diff
@@ -0,0 +1,739 @@
...
+			}
+
+			// Wait for the informer to get the event.
+			select {
+			case <-certManagerChan:
+			case <-time.After(wait.ForeverTestTimeout):
+				t.Fatal("Informer did not get the added cert manager object")
+			}
+
+			err = withContainerResourcesValidateHook(certManagerInformers, tc.deploymentName)(nil, nil)
```
[the remark this answers is not in the document]
Comment (thejasn) [reply pkg/controller/deployment/deployment_overrides_validation_test.go:718]: We can ditch this in favour of covering this as part of e2e.

**Review by thejasn — CHANGES_REQUESTED** [review]

File: pkg/controller/deployment/deployment_overrides_validation_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_overrides_validation_test.go:718]: > @arkadeepsen I believe we are not really testing anything with this function correct me if I'm mistaken. Since you passing nil as args to the returned function which doesn't use any of the arguments in the returned function. I don't see the value here.

The function which is returned doesn't use any of the arguments. Thus passing anything in the call wouldn't make any difference. Also the `withContainerResourcesValidateHook` function is similar to the existing `ValidateHook` functions. 

> We can ditch this in favour of covering this as part of e2e.

The unit test is added to test whether this function is working as expected. We are anyway testing the overall functionality through the e2e test.

File: pkg/controller/deployment/deployment_overrides_test.go (hunk 9 above)
Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_overrides_test.go:430]: This unit test is for checking whether the `withContainerResourcesOverrideHook` is working correctly or not with the inputs provided. The e2e test will anyway test the overall functionality of the feature.

File: pkg/controller/deployment/deployment_overrides_validation_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (thejasn) [reply pkg/controller/deployment/deployment_overrides_validation_test.go:718]: > The unit test is added to test whether this function is working as expected.

Yes, but if what is the critical business logic that you are testing here? All the functions that are constructed inside `withContainerResourcesValidateHook()` are not dependent on the args and is only dependent on the cluster object. `validateResources` is the actual function you we want to test (and this function is pretty much static if we remove the dep on the informer), I'd suggest to update the scope and make it testable directly. WDYT? This would reduce the boiler plate required for tests and also tests the critical bits.

[the remark this answers is not in the document]
Comment (swghosh) [reply pkg/controller/deployment/deployment_overrides_validation_test.go:718]: please refer to my other comment on https://github.com/openshift/cert-manager-operator/pull/136#discussion_r1285910160, ditto.

File: pkg/controller/deployment/deploment_helper_test.go [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+					corev1.ResourceMemory: k8sresource.MustParse("64Mi"),
+				},
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    k8sresource.MustParse("40m"),
+					corev1.ResourceMemory: k8sresource.MustParse("64Mi"),
+				},
+			},
+		},
+		{
+			name: "override resources limits and requests merges with source resource limits and requests respectively",
```
Comment (swghosh) [inline pkg/controller/deployment/deploment_helper_test.go:177]: small nit: would you mind chaning to:
"override resources replace source limits and requests" instead?

Comment (swghosh) [reply pkg/controller/deployment/deploment_helper_test.go:177]:
```suggestion
			name: "override resources replace source limits and requests",
```

File: pkg/controller/deployment/deployment_overrides_validation_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_overrides_validation_test.go:718]: Ok. I'll change the scope of `validateResources` and add unit test for it. I'll remove the unit test for `withContainerResourcesValidateHook`.

File: pkg/controller/deployment/deployment_overrides_test.go (hunk 9 above)
Comment (arkadeepsen) [reply pkg/controller/deployment/deployment_overrides_test.go:430]: Ok. I'll remove the unit test for `withContainerResourcesOverrideHook`.

File: pkg/controller/deployment/deploment_helper_test.go [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			actualResources := mergeContainerResources(tc.sourceResources, tc.overrideResources)
+			require.Equal(t, tc.expected, actualResources)
+		})
+	}
+}
+
+func TestGetOverrideResourcesFor(t *testing.T) {
```
Comment (arkadeepsen) [inline pkg/controller/deployment/deploment_helper_test.go:215]: @thejasn @swghosh PLMK if this unit test is also redundant. I'll remove this one as well.

File: pkg/controller/deployment/deploment_helper_test.go (hunk 11 above)
Comment (arkadeepsen) [reply pkg/controller/deployment/deploment_helper_test.go:177]: This test actually merges. There's not replace of existing values.

File: pkg/controller/deployment/deploment_helper_test.go (hunk 12 above)
Comment (swghosh) [reply pkg/controller/deployment/deploment_helper_test.go:215]: I think we shouldn’t remove this one, but rather keep it. Cause, it effectively tests the switch case in the actual func that switches between different deployments (in the past we’ve seen a bug where we accidentally made mistakes in such For functions, so better to retain this).

File: pkg/controller/deployment/deploment_helper_test.go (hunk 11 above)
Comment (swghosh) [reply pkg/controller/deployment/deploment_helper_test.go:177]: Cool, then leave it as is.

**Comment by swghosh** [comment]
/cc @davemulford 
for px-approved label
/cc @xingxingxia 
for qe-approved label
(you should be able to test this once we do our first build for 1.12, post-merge test cause this is dev-tested)
/cc @xenolinux 
for docs-approved label

TIA all!

**Comment by thejasn** [comment]
/label px-approved
Adding labels since this is blocking 1.12 release candidate build.

**Comment by xingxingxia** [comment]
/label qe-approved
Glad to see test/e2e/cert_manager_deployment_test.go already covers "cert-manager" deployment. QE decide to do post-merge test for cert-manager-cainjector, and the cert-manager-webhook and operator.
